In [1]:
import pandas as pd

In [2]:
df = pd.read_csv(r"D:\DELL G5\DA\stack-overflow-developer-survey-2025\survey_results_public.csv", low_memory=False)
sch = pd.read_csv(r"D:\DELL G5\DA\stack-overflow-developer-survey-2025\survey_results_schema.csv", low_memory=False)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49191 entries, 0 to 49190
Columns: 172 entries, ResponseId to JobSat
dtypes: float64(52), int64(1), object(119)
memory usage: 64.6+ MB


In [4]:
pd.set_option('display.max_columns', None)

In [5]:
df.head()

ResponseId                      MainBranch              Age  \
0           1  I am a developer by profession  25-34 years old   
1           2  I am a developer by profession  25-34 years old   
2           3  I am a developer by profession  35-44 years old   
3           4  I am a developer by profession  35-44 years old   
4           5  I am a developer by profession  35-44 years old   

                                           EdLevel  \
0  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
1              Associate degree (A.A., A.S., etc.)   
2     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   

                                          Employment  \
0                                           Employed   
1                                           Employed   
2  Independent contractor, freelancer, or self-em...   
3                                           Employed   
4  Independent contractor, freelancer, or self-em...   

                                    EmploymentAddl  WorkExp  \
0  Caring for dependents (children, elderly, etc.)      8.0   
1                                              NaN      2.0   
2                                None of the above     10.0   
3                                None of the above      4.0   
4  Caring for dependents (children, elderly, etc.)     21.0   

                                     LearnCodeChoose  \
0  Yes, I am not new to coding but am learning ne...   
1  Yes, I am not new to coding but am learning ne...   
2  Yes, I am not new to coding but am learning ne...   
3  Yes, I am not new to coding but am learning ne...   
4  No, I am not new to coding and did not learn n...   

                                           LearnCode  \
0  Online Courses or Certification (includes all ...   
1  Online Courses or Certification (includes all ...   
2  Online Courses or Certification (includes all ...   
3  Other online resources (e.g. standard search, ...   
4                                                NaN   

                                         LearnCodeAI  \
0  Yes, I learned how to use AI-enabled tools for...   
1  Yes, I learned how to use AI-enabled tools for...   
2  Yes, I learned how to use AI-enabled tools for...   
3  Yes, I learned how to use AI-enabled tools for...   
4  Yes, I learned how to use AI-enabled tools for...   

                                          AILearnHow  YearsCode  \
0                AI CodeGen tools or AI-enabled apps       14.0   
1                AI CodeGen tools or AI-enabled apps       10.0   
2  AI CodeGen tools or AI-enabled apps;Technical ...       12.0   
3  AI CodeGen tools or AI-enabled apps;Videos (no...        5.0   
4  Technical documentation (is generated for/by t...       22.0   

                DevType                   OrgSize                  ICorPM  \
0     Developer, mobile        20 to 99 employees          People manager   
1   Developer, back-end      500 to 999 employees  Individual contributor   
2  Developer, front-end                       NaN                     NaN   
3   Developer, back-end  10,000 or more employees  Individual contributor   
4   Engineering manager                       NaN                     NaN   

                                          RemoteWork  \
0                                             Remote   
1  Hybrid (some in-person, leans heavy to flexibi...   
2                                                NaN   
3                                             Remote   
4                                                NaN   

                                   PurchaseInfluence  TechEndorseIntro  \
0  Yes, I influenced the purchase of a substantia...              Work   
1                                                 No  Personal Project   
2                                                 No              Work   
3                                                 No  P

In [6]:
sch.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 139 entries, 0 to 138
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   qid       139 non-null    object 
 1   qname     139 non-null    object 
 2   question  139 non-null    object 
 3   type      139 non-null    object 
 4   sub       49 non-null     object 
 5   sq_id     49 non-null     float64
dtypes: float64(1), object(5)
memory usage: 6.6+ KB


In [7]:
pd.set_option('display.max_colwidth', None)

In [8]:
sch.head()

qid          qname  \
0  QID18  TechEndorse_1   
1  QID18  TechEndorse_2   
2  QID18  TechEndorse_3   
3  QID18  TechEndorse_4   
4  QID18  TechEndorse_5   

                                                                                   question  \
0  What attracts you to a technology or causes you to endorse it (most to least important)?   
1  What attracts you to a technology or causes you to endorse it (most to least important)?   
2  What attracts you to a technology or causes you to endorse it (most to least important)?   
3  What attracts you to a technology or causes you to endorse it (most to least important)?   
4  What attracts you to a technology or causes you to endorse it (most to least important)?   

  type                                      sub  sq_id  
0   RO  AI integration or AI Agent capabilities    1.0  
1   RO                          Easy-to-use API    2.0  
2   RO                  Robust and complete API    3.0  
3   RO     Customizable and manageable codebase    4.0  
4   RO                   Reputation for quality    5.0

1. Підрахунок загальної кількості респондентів

In [9]:
all_resp = df['ResponseId'].nunique()
all_resp

49191

У опитуванні брало участь 49191 респондентів

2. Аналіз повноти відповідей респондентів

In [10]:
questions = sch['qname'].dropna().unique()

In [11]:
que_col = set(questions).intersection(set(df.columns))

In [12]:
df_questions = df[list(que_col)]

In [13]:
all_que_response = df_questions.dropna()

In [14]:
all_que_response.shape[0]

0

In [15]:
#кількість питань на які мали відповісти респонденти
len(que_col)

126

Висновок: Жоден із респондентів не відповів на абсолютно всі питання опитування. Це пояснюється великою кількістю запитань (126), наявністю необов’язкових полів та логікою анкети, за якої частина питань стає неактуальною залежно від попередніх відповідей

3. Статистичний аналіз досвіду респондентів

In [16]:
df['WorkExp'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 49191 entries, 0 to 49190
Series name: WorkExp
Non-Null Count  Dtype  
--------------  -----  
42893 non-null  float64
dtypes: float64(1)
memory usage: 384.4 KB


In [17]:
df['WorkExp'].isna().sum()

np.int64(6298)

In [18]:
df['WorkExp'].dropna()

0         8.0
1         2.0
2        10.0
3         4.0
4        21.0
         ... 
49183     2.0
49184    30.0
49187    25.0
49188    17.0
49189     2.0
Name: WorkExp, Length: 42893, dtype: float64

In [19]:
print (f'Відповідь:')
print (f'Міри центральної тенденції:')
print (f'Середнє значення: {round(df['WorkExp'].mean(), 2)}'),
print (f'Медіана: {df['WorkExp'].median()}'),
print (f'Мода: {df['WorkExp'].mode()[0]}')

Відповідь:
Міри центральної тенденції:
Середнє значення: 13.37
Медіана: 10.0
Мода: 10.0


4. Аналіз віддаленої роботи

In [20]:
sch[sch['qname'].str.contains('remote', case=False, na=False)]

qid       qname                                           question type  \
63  QID24  RemoteWork  Which best describes your current work situation?   MC   

    sub  sq_id  
63  NaN    NaN

In [21]:
df['RemoteWork'].unique()

array(['Remote', 'Hybrid (some in-person, leans heavy to flexibility)',
       nan, 'In-person', 'Hybrid (some remote, leans heavy to in-person)',
       'Your choice (very flexible, you can come in when you want or just as needed)'],
      dtype=object)

In [22]:
count_remote = df[df['RemoteWork'] == 'Remote'].shape[0]
count_remote

10931

Висновок: Із опитуваних віддалено працюють 10931 респондентів

5. Визначення популярності Python

In [23]:
[col for col in df.columns if 'Language' in col]

['LanguageChoice',
 'LanguageHaveWorkedWith',
 'LanguageWantToWorkWith',
 'LanguageAdmired',
 'LanguagesHaveEntry',
 'LanguagesWantEntry']

In [24]:
df['LanguageHaveWorkedWith'].head(10)

0                                                                         Bash/Shell (all shells);Dart;SQL
1                                                                                                     Java
2                                                                      Dart;HTML/CSS;JavaScript;TypeScript
3                                                                                          Java;Kotlin;SQL
4    C;C#;C++;Delphi;HTML/CSS;Java;JavaScript;Lua;PowerShell;Python;SQL;TypeScript;VBA;Visual Basic (.Net)
5                                                                                               Java;Scala
6                                                                                    JavaScript;TypeScript
7                                        Bash/Shell (all shells);HTML/CSS;JavaScript;Python;SQL;TypeScript
8                                                                                        Java;Python;Scala
9                                    

In [25]:
use_python = df['LanguageHaveWorkedWith'].str.contains('Python', na=False)

In [26]:
python_users = use_python.sum()

In [27]:
python_per = round(python_users/all_resp*100, 2)
python_per

np.float64(37.54)

In [28]:
print('Висновок: Відсоток респондентів, які програмують на Python:', python_per, '%')

Висновок: Відсоток респондентів, які програмують на Python: 37.54 %


6. Аналіз шляхів навчання програмуванню

In [29]:
sch[sch['qname'].str.contains('learn', case=False, na=False)]

qid            qname  \
55  QID10  LearnCodeChoose   
56  QID11        LearnCode   
57  QID12      LearnCodeAI   
58  QID32       AILearnHow   

                                                                                                     question  \
55          Did you begin learning to code or learn a new coding skill/programming language in the last year?   
56                                         How did you learn to code in the past year? Select all that apply.   
57  Did you spend time in the last year learning AI programming or AI-enabled tooling on your own or at work?   
58                                  How did you learn to code for AI in the past year? Select all that apply.   

   type  sub  sq_id  
55   MC  NaN    NaN  
56   MC  NaN    NaN  
57   MC  NaN    NaN  
58   MC  NaN    NaN

In [30]:
df['LearnCode'].head()

0                                                                                                                                                    Online Courses or Certification (includes all media types);Other online resources (e.g. standard search, forum, online community)
1                       Online Courses or Certification (includes all media types);Other online resources (e.g. standard search, forum, online community);Books / Physical media;Videos (not associated with specific online course or certification);Stack Overflow or Stack Exchange
2                                                                                     Online Courses or Certification (includes all media types);Videos (not associated with specific online course or certification);Technical documentation (is generated for/by the tool or system)
3    Other online resources (e.g. standard search, forum, online community);Videos (not associated with specific online course or certification);Stack Overflow or 

In [31]:
online_courses = df['LearnCode'].str.contains('Online Courses', na=False)

In [32]:
online_courses.sum()

np.int64(10973)

Висновок: 10973 респондентів навчалося програмувати за допомогою онлайн курсів

7. Географічний аналіз компенсації Python-розробників

In [33]:
python_df = df[df['LanguageHaveWorkedWith'].str.contains('Python', na=False)]

In [34]:
python_df = python_df[['Country', 'ConvertedCompYearly']]
python_df

Country  \
4                                                   Ukraine   
7                                                   Ukraine   
8                                                   Ukraine   
15                                                    India   
17                                                   Greece   
...                                                     ...   
49143  United Kingdom of Great Britain and Northern Ireland   
49145                              United States of America   
49175                                               Germany   
49176                                                 Kenya   
49177                              United States of America   

       ConvertedCompYearly  
4                  60000.0  
7                  72000.0  
8                  70000.0  
15                     NaN  
17                     NaN  
...                    ...  
49143                  NaN  
49145                  NaN  
49175                  NaN  
49176                  NaN  
49177                  NaN  

[18466 rows x 2 columns]

In [35]:
python_df = python_df.dropna(subset=['ConvertedCompYearly'])
python_df

Country  ConvertedCompYearly
4                       Ukraine              60000.0
7                       Ukraine              72000.0
8                       Ukraine              70000.0
22     United States of America              87500.0
27                       Sweden             209643.0
...                         ...                  ...
49028  United States of America              77000.0
49088  United States of America             280000.0
49089                     India               4650.0
49105                     Egypt                719.0
49106              South Africa              69008.0

[12765 rows x 2 columns]

In [36]:
comp_by_country = python_df.groupby('Country')['ConvertedCompYearly'].agg(['mean', 'median']).reset_index().round(1)

In [37]:
comp_by_country.columns = ['Country', 'AvgCompensation', 'MedianCompensation']

In [38]:
comp_by_country

Country  AvgCompensation  MedianCompensation
0                             Afghanistan          22328.7              1000.0
1                                 Albania          47217.6             50000.0
2                                 Algeria          20187.3              7088.0
3                                 Andorra         226103.5            226103.5
4                     Antigua and Barbuda              1.0                 1.0
..                                    ...              ...                 ...
148  Venezuela, Bolivarian Republic of...           9908.6              3000.0
149                              Viet Nam         218837.2              8254.0
150                                 Yemen          32929.5             23672.0
151                                Zambia           5424.2              3206.0
152                              Zimbabwe          34000.0             25500.0

[153 rows x 3 columns]

8. Аналіз освіти найбільш оплачуваних спеціалістів

In [39]:
ed_comp = df[['EdLevel', 'ConvertedCompYearly']]

In [40]:
ed_comp = ed_comp.dropna(subset=['ConvertedCompYearly'])
ed_comp

EdLevel  \
0             Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
1                         Associate degree (A.A., A.S., etc.)   
2                Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3                Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4             Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
...                                                       ...   
49098          Professional degree (JD, MD, Ph.D, Ed.D, etc.)   
49105  Some college/university study without earning a degree   
49106            Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
49121                     Associate degree (A.A., A.S., etc.)   
49122  Some college/university study without earning a degree   

       ConvertedCompYearly  
0                  61256.0  
1                 104413.0  
2                  53061.0  
3                  36197.0  
4                  60000.0  
...                    ...  
49098               2729.0  
49105                719.0  
49106              69008.0  
49121              58007.0  
49122              46406.0  

[23947 rows x 2 columns]

In [41]:
ed_comp = ed_comp.sort_values(by='ConvertedCompYearly', ascending=False)

In [42]:
top5 = ed_comp.head()
top5

EdLevel  ConvertedCompYearly
34267              Associate degree (A.A., A.S., etc.)           50000000.0
28700  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)           33552715.0
43143              Associate degree (A.A., A.S., etc.)           18387548.0
35353     Bachelor’s degree (B.A., B.S., B.Eng., etc.)           15430267.0
45971  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)           13921760.0

In [43]:
top5[['EdLevel']]

EdLevel
34267              Associate degree (A.A., A.S., etc.)
28700  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)
43143              Associate degree (A.A., A.S., etc.)
35353     Bachelor’s degree (B.A., B.S., B.Eng., etc.)
45971  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)

Висновок: Серед п’яти респондентів з найвищою компенсацією всі мають вищу освіту різного рівня

9. Аналіз популярності Python по віковим категоріям 

In [44]:
df['UsesPython'] = use_python

In [45]:
age_python = df.groupby('Age')['UsesPython'].mean().reset_index()
age_python

Age  UsesPython
0    18-24 years old    0.400000
1    25-34 years old    0.369393
2    35-44 years old    0.367193
3    45-54 years old    0.386295
4    55-64 years old    0.372430
5  65 years or older    0.316348
6  Prefer not to say    0.312169

In [46]:
age_python['PythonPopular'] = (age_python['UsesPython']*100).round(2)

In [47]:
age_python = age_python[['Age', 'PythonPopular']].sort_values(by='PythonPopular', ascending=False)
age_python

Age  PythonPopular
0    18-24 years old          40.00
3    45-54 years old          38.63
4    55-64 years old          37.24
1    25-34 years old          36.94
2    35-44 years old          36.72
5  65 years or older          31.63
6  Prefer not to say          31.22

Висновок: Python є найбільш популярним серед респондентів вікової категорії 18–24 років

10. Аналіз індустрій серед високооплачуваних віддалених працівників

In [48]:
df_for_ind = df[['ConvertedCompYearly', 'RemoteWork', 'Industry']]
df_for_ind

ConvertedCompYearly  \
0                  61256.0   
1                 104413.0   
2                  53061.0   
3                  36197.0   
4                  60000.0   
...                    ...   
49186                  NaN   
49187                  NaN   
49188                  NaN   
49189                  NaN   
49190                  NaN   

                                                RemoteWork  \
0                                                   Remote   
1      Hybrid (some in-person, leans heavy to flexibility)   
2                                                      NaN   
3                                                   Remote   
4                                                      NaN   
...                                                    ...   
49186                                                  NaN   
49187                                                  NaN   
49188  Hybrid (some in-person, leans heavy to flexibility)   
49189                                                  NaN   
49190                                                  NaN   

                           Industry  
0                           Fintech  
1      Retail and Consumer Services  
2              Software Development  
3      Retail and Consumer Services  
4              Software Development  
...                             ...  
49186                           NaN  
49187                           NaN  
49188    Banking/Financial Services  
49189                           NaN  
49190                           NaN  

[49191 rows x 3 columns]

In [49]:
df_for_ind = df_for_ind.dropna(subset=['ConvertedCompYearly'])

In [50]:
perc_75 = df_for_ind['ConvertedCompYearly'].quantile(0.75)
perc_75

np.float64(120596.0)

In [51]:
high_paid_remote = df_for_ind[(df_for_ind['ConvertedCompYearly'] > perc_75) & (df_for_ind['RemoteWork'] == 'Remote')]

In [52]:
high_paid_remote

ConvertedCompYearly RemoteWork                      Industry
29                205000.0     Remote          Software Development
31                250000.0     Remote                       Fintech
41                270000.0     Remote          Software Development
42                250000.0     Remote          Software Development
48                142949.0     Remote                       Fintech
...                    ...        ...                           ...
48913             129334.0     Remote          Software Development
48937             124738.0     Remote  Retail and Consumer Services
48970             350000.0     Remote                    Healthcare
49083             188000.0     Remote  Retail and Consumer Services
49088             280000.0     Remote          Software Development

[2461 rows x 3 columns]

In [53]:
industry = high_paid_remote['Industry'].value_counts().reset_index()
industry

Industry  count
0                         Software Development   1186
1                                      Fintech    190
2                                   Healthcare    188
3                                       Other:    176
4   Internet, Telecomm or Information Services    138
5                   Banking/Financial Services     88
6                                   Government     78
7                 Media & Advertising Services     75
8                 Retail and Consumer Services     65
9              Transportation, or Supply Chain     63
10        Computer Systems Design and Services     59
11                                   Insurance     46
12                                      Energy     45
13                               Manufacturing     33
14                            Higher Education     30

Висновок: Серед високооплачуваних респондентів, які працюють віддалено, найпоширенішими індустріями є Software Development, Fintech та Healthcare